Source/VMware.vSphereDSC/DSCResources/VMHostNetwork/VMHostVirtualPortGroup.ps1

```diff
@@ -0,0 +1,236 @@
+
+        return $result
+    }
+
+    <#
+    .DESCRIPTION
+
+    Retrieves the Virtual Port Group with the specified Name, available on the specified Virtual Switch and VMHost from the server if it exists,
+    otherwise returns $null.
+    #>
+    [PSObject] GetVirtualPortGroup($vmHost, $foundVirtualSwitch) {
+        return Get-VirtualPortGroup -Server $this.Connection -Name $this.PortGroupName -VirtualSwitch $foundVirtualSwitch -VMHost $vmHost -ErrorAction SilentlyContinue
+    }
+
+    <#
+    .DESCRIPTION
+
+    Ensures that the passed VLanId value is in the range [0, 4095].
+    #>
+    [void] EnsureVLanIdValueIsValid() {
+        if ($this.VLanId -lt 0 -or $this.VLanId -gt $this.VLanIdMaxValue) {
+            throw "The passed VLanId value $($this.VLanId) is not valid. The valid values are in the following range: [0, $($this.VLanIdMaxValue)]."
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Checks if the VLanId is specified and needs to be updated.
+    #>
+    [bool] ShouldUpdateVirtualPortGroup($portGroup) {
+        if ($null -eq $this.VLanId) {
+            return $false
+        }
+
+        return ($this.VLanId -ne $portGroup.VLanId)
+    }
+
+    <#
+    .DESCRIPTION
+
+    Returns the populated Port Group parameters.
+    #>
+    [hashtable] GetPortGroupParams() {
+        $portGroupParams = @{}
+
+        $portGroupParams.Confirm = $false
+        $portGroupParams.ErrorAction = 'Stop'
+
+        if ($null -ne $this.VLanId) {
+            $this.EnsureVLanIdValueIsValid()
+            $portGroupParams.VLanId = $this.VLanId
+        }
+
+        return $portGroupParams
+    }
+
+    <#
+    .DESCRIPTION
+
+    Creates a new Port Group available on the specified Virtual Switch.
+    #>
+    [void] AddVirtualPortGroup($foundVirtualSwitch) {
+        $portGroupParams = $this.GetPortGroupParams()
+
+        $portGroupParams.Server = $this.Connection
+        $portGroupParams.Name = $this.PortGroupName
+        $portGroupParams.VirtualSwitch = $foundVirtualSwitch
+
+        try {
+            New-VirtualPortGroup @portGroupParams
+        }
+        catch {
+            throw "Cannot create Virtual Port Group $($this.PortGroupName). For more information: $($_.Exception.Message)"
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Updates the Port Group by changing its VLanId value.
+    #>
+    [void] UpdateVirtualPortGroup($portGroup) {
+        $portGroupParams = $this.GetPortGroupParams()
+
+        try {
+            $portGroup | Set-VirtualPortGroup @portGroupParams
+        }
+        catch {
+            throw "Cannot update Virtual Port Group $($this.PortGroupName). For more information: $($_.Exception.Message)"
+        }
+    }
+
+    <#
+    .DESCRIPTION
+
+    Removes the specified Port Group available on the Virtual Switch.
+    #>
+    [void] RemoveVirtualPortGroup($portGroup) {
+        try {
+            $portGroup | Remove-VirtualPortGroup -Confirm:$false -ErrorAction Stop
```

Comment:
What happens with the VMs connected to the portgroup when Remove-VirtualPortGroup is called? Does the remove cmdlet fail?

Reply:
If all VMs are PoweredOff the cmdlet succeeds and the Port Group is deleted. However if one of the VMs is PoweredOn, the cmdlet fails with the following error: 'The resource '<port group name>' is in use.' and the Port Group is not deleted. I will describe this case in the documentation.